courses/utils.py

```diff
@@ -54,7 +56,16 @@ def generate_program_certificate(user, program):
         )
         return existing_cert_queryset.first(), False
 
-    courses_in_program_ids = set(program.courses.values_list("id", flat=True))
+    courses_in_program_ids = set(
+        program.get_requirements_root()
+        .get_children()
+        .filter(operator=ProgramRequirement.Operator.ALL_OF)
+        .first()
+        .get_children()
+        .filter(node_type=ProgramRequirementNodeType.COURSE)
+        .values_list("course_id", flat=True)
+    )
```

Comment:
So it won't be as straightforward as just changing this query. I think what your want to do is:

- Do this query, although you need to modify it a bit both to simplify it and also handle `MIN_NUMBER_OF` operators:
```suggestion
    courses_in_program_ids = set(
        program.get_requirements_root()
        .get_descendants()
        .filter(node_type=ProgramRequirementNodeType.COURSE)
        .values_list("course_id", flat=True)
    )
```
Then, you'll need to walk the tree of requirements checking that against the course certificates. I wrote some pseudocode for this [here](https://github.com/mitodl/mitxonline/pull/1070#issuecomment-1262632406) a while back, although it incorrectly uses grades.

Reply:
Added `MIN_NUMBER_OF` check and updated query, please have a look